blocks/http/test/qa_HttpBlock.cpp

```diff
@@ -0,0 +1,215 @@
+#include <future>
+
+#include <boost/ut.hpp>
+
+#include <fmt/format.h>
+#include <fmt/ranges.h>
+
+#include <gnuradio-4.0/basic/common_blocks.hpp>
+#include <gnuradio-4.0/Block.hpp>
+#include <gnuradio-4.0/Buffer.hpp>
+#include <gnuradio-4.0/Graph.hpp>
+#include <gnuradio-4.0/reflection.hpp>
+#include <gnuradio-4.0/Scheduler.hpp>
+
+#include <gnuradio-4.0/http/HttpBlock.hpp>
+
+template<typename T>
+struct fixed_source : public gr::Block<fixed_source<T>, gr::PortOutNamed<T, "out">> {
+    using super_t = gr::Block<fixed_source<T>, gr::PortOutNamed<T, "out">>;
+    T value = 1;
+
+    gr::work::Result
+    work(std::size_t requested_work) {
+        auto &port   = gr::outputPort<0, gr::PortType::STREAM>(this);
+        auto &writer = port.streamWriter();
+        auto  data   = writer.reserve_output_range(1UZ);
+        data[0]      = value;
+        data.publish(1UZ);
+
+        value += 1;
+        return { requested_work, 1UZ, gr::work::Status::OK };
+    }
+
+    void
+    trigger() {
+        gr::Message message;
+        message[gr::message::key::Kind] = "custom_kind";
+        super_t::emitMessage(super_t::msgOut, message);
+    }
+};
+
+const boost::ut::suite HttpBlocktests = [] {
+    using namespace boost::ut;
+    using namespace gr;
+    using namespace std::literals;
+    using namespace std::chrono_literals;
+    "http GET"_test = [] {
+#ifdef __EMSCRIPTEN__
+        std::thread thread{ [&]() {
+            // see ./pre.js for the emscripten server implementation
+            emscripten_run_script("startServer();");
+        } };
+#else
+        httplib::Server server;
+        server.Get("/echo", [](const httplib::Request, httplib::Response &res) { res.set_content("Hello world!", "text/plain"); });
+
+        auto thread = std::thread{ [&server] { server.listen("localhost", 8080); } };
+        server.wait_until_ready();
+#endif
+
+        gr::Graph graph;
+        auto     &source    = graph.emplaceBlock<fixed_source<uint8_t>>();
+        auto     &httpBlock = graph.emplaceBlock<http::HttpBlock<uint8_t>>("http://localhost:8080", "/echo");
+        auto     &sink      = graph.emplaceBlock<InspectSink<pmtv::map_t>>();
+        graph.connect(source, 0, httpBlock, 0);
+        std::ignore = source.msgOut.connect(httpBlock.msgIn);
+        graph.connect(httpBlock, 0, sink, 0);
+        httpBlock.start();
+
+        // make a request
+        source.trigger();
+        httpBlock.processScheduledMessages();
+        pmtv::map_t response;
+        while (response.empty()) {
+            std::this_thread::sleep_for(1ms);
+            std::ignore = source.work(1UZ);
+            auto work   = httpBlock.work(1UZ);
+            work        = sink.work(1UZ);
+            response    = sink.value;
+        }
```

Comment:
I saw that you use while loops for all unit tests to run the graph. Will it work with the standard `Scheduler`? Why don't you use it?

Reply:
Yes it should just work with a scheduler.
The reason that I don't use it here, is because it allows me to have more control over the execution in the graph. 
Furthermore it means I don't have to rely on the graph execution to stop automatically due to forward propagation of blocks turning into the `STOPPED` state.
Once #244 is merged, I might reconsider using the scheduler for tests though ;)

Reply:
I strongly recommend and request to use the scheduler interface, as this ensures that the block is ready for production use, which invariably relies on a scheduler. We've already encountered scenarios, such as with the `Selector` block, where its bespoke `work()` iteration method in the unit-tests falters in real-world applications. Moreover, the re-implementation of a block's low-level `work()` function is, or ought to be, strongly discouraged, given that many users fail to implement other critical invariants such as the state machine, tag propagation, sampling ratio, etc.

It's pertinent to note that there isn't a singular scheduler implementation; rather, the aim is for users to feel empowered and encouraged to develop their own. Nonetheless, for unit tests, it's imperative that the blocks at least function with the `gr::scheduler::Simple` implementation, which stands as one of the two compliant implementations.

Given the reasons outlined above, it's improbable that we'll proceed with merging [#244](https://github.com/fair-acc/graph-prototype/pull/244), as it represents a workaround and overlooks many of the invariants that both the scheduler and block implementations are required to fulfil.

Reply:
Do I understand it correctly that the main problem now which prevent you from usage of `Scheduler` is that you can not stop scheduler execution? The execution is stopped using `END_OF_STREAM` tag, which should be publish by source. In your case if you use `Scheduler`, it will run forever because `EOF` tag is not published.

Here is one of the simple example of source:
```cpp
template<typename T>
struct CountSource : public gr::Block<CountSource<T>> {
    gr::PortOut<T> out{};
    int            count{ 0 };
    int            nSamples{ 1024 };

    constexpr T
    processOne() {
        count++;
        if (count >= nSamples) {
            this->requestStop(); 
        }
        return static_cast<T>(count - 1); // -1 to start from 0
    }
};
```

Reply:
Expanding on the previous discussion:

Regarding the unit tests: the current implementation involves the trigger mechanism using input port of `HttpBlock`. One should use `ClockSource` block or a block with similar functionality then as input. This would allow us to define a specific time frame, after which an `EOS` tag is published. Specifically, in the case of `ClockSource`, this could be effectively managed by setting parameters like `n_samples_max` and `sample_rate`. Integrating a `Scheduler` would ensure the process halts upon reaching the predefined time limit.

This leads to a broader question: Is an input port absolutely necessary for `HttpBlock`? Alternatively, could we design the `HttpBlock` to function autonomously as a source block, issuing the `EOS` tag independently? Start/Stop mechanism can be then done via message port.

Reply:
@vimpostor, do you have any updates? After reviewing your pull request again and discussing it with Ralph, we have some suggestions on how it might be integrated with the Scheduler. Essentially, you could implement a sink that stops the scheduler under certain conditions. Here's an outline of the code approach:

Firstly, you'll need to create a custom sink block as follows:
```cpp
template<typename T>
class HttpTestSink : public gr::Block<HttpTestSink<T>> {
public:
    gr::PortIn<T>         in;
    T                     value{};
    std::function<void()> stopFunc;

    constexpr void
    processOne(T val) {
        if (val.empty()) {
            stopFunc();
        } else {
            value = val;
        };
    }
};

ENABLE_REFLECTION_FOR_TEMPLATE_FULL((typename T), (HttpTestSink<T>), in);
```
Then, set up the Graph, Scheduler, and execute it as shown below:
```cpp
gr::Graph graph;
auto     &source    = graph.emplaceBlock<fixed_source<uint8_t>>();
auto     &httpBlock = graph.emplaceBlock<http::HttpBlock<uint8_t>>("http://localhost:8080", "/echo");
auto     &sink      = graph.emplaceBlock<HttpTestSink<pmtv::map_t>>();

expect(eq(ConnectionResult::SUCCESS, graph.connect<"out">(source).template to<"in">(httpBlock)));
expect(eq(ConnectionResult::SUCCESS, source.msgOut.connect(httpBlock.msgIn)));
expect(eq(ConnectionResult::SUCCESS, graph.connect<"out">(httpBlock).template to<"in">(sink)));

auto sched    = gr::scheduler::Simple<>(std::move(graph), threadPool);
sink.stopFunc = [&sched]() { sched.stop(); };

source.trigger();
httpBlock.processScheduledMessages();
sched.runAndWait();
pmtv::map_t response = sink.value;
expect(eq(std::get<std::string>(sink.value.at("raw-data")), "Hello world!"sv));
```


Also as I wrote in the comment to the code. You have to remove these code from `Http::startThread()` because `_state` is always `RUNNING` at this point:
```cpp
lifecycle::State expectedThreadState = lifecycle::State::INITIALISED;
if (!this->_state.compare_exchange_strong(expectedThreadState, lifecycle::State::RUNNING, 
   std::memory_order_acq_rel)) {
   return;
}
this->_state.notify_all();
```

I've tested this code on my machine, and the test passed successfully.

Alternative Approach. As previously mentioned, another option could involve creating a Source with a timeout mechanism, namely send EOS tag after timeout.

I hope this clarifies our suggestions. Looking forward to your thoughts and any further updates you might have.